File: LC155_MinStack/Stack/MinStack.java
import java.util.Stack;

class MinStack {
    Stack<Integer> s;
    Stack<Integer> minS;

    public MinStack() {
        this.s = new Stack<>();
        this.minS = new Stack<>();
    }
    
    public void push(int val) {
        // Time Complexity: O(1)
        s.push(val);
        if(minS.isEmpty() || minS.peek() >= val) minS.push(val);       
    }
    
    public void pop() {
        // Time Complexity: O(1)
        int temp = s.pop();
        if(minS.peek() == temp) minS.pop();
    }
    
    public int top() {
        // Time Complexity: O(1)
        return s.peek();
    }
    
    public int getMin() {
        // Time Complexity: O(1)
        return minS.peek();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: LC155_MinStack/Stack/MinStack.js

var MinStack = function() {
    this.stack = []
    this.minStack = []
};

/** 
 * @param {number} val
 * @return {void}
 */
MinStack.prototype.push = function(val) {
    // Time Complexity: O(1)
    this.stack.push(val)
    if(this.minStack.length === 0 || this.minStack[this.minStack.length-1] >= val){
        this.minStack.push(val)
    }
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
    // Time Complexity: O(1)
    let retVal = this.stack.pop()
    if(retVal === this.minStack[this.minStack.length-1]){
        this.minStack.pop()
    }
    return retVal
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    // Time Complexity: O(1)
    return this.stack[this.stack.length-1]
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
    // Time Complexity: O(1)
    return this.minStack[this.minStack.length-1]
};

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */